### PR TITLE
(WIP) (FM-2802) Introduce global config and credentials files in Puppet confdir

### PR DESCRIPTION
**Pull request by garethr**
This is to work around the Puppet agent nuking environment variables,
meaning the default location (based on HOME) doesn't work correctly.
This also adds support for storing the other settings in global files.

### CODE REVIEW

**Comment by daveseff** [comment]
#thumbsup

**Comment by esalberg** [comment]
We forked this change, and it's working for us, so merging it in would be great.

**Comment by aharden** [comment]
Consider this another vote for getting this merged in ASAP.  I've been running off a fork of this module for months to be able to use our http proxy for AWS API access during Puppet agent runs.

File: lib/puppet_x/puppetlabs/aws.rb [hunk 1]
```diff
@@ -56,21 +56,85 @@ def self.read_only(*methods)
...
 
       def self.client_config(region)
-        config = {region: region, logger: logger}
-        if ENV['PUPPET_AWS_PROXY'] and not ENV['PUPPET_AWS_PROXY'].empty?
-          config[:http_proxy] = ENV['PUPPET_AWS_PROXY']
+        config = {logger: logger}
+        config[:http_proxy] = proxy_configuration if proxy_configuration
+        config[:credentials] = global_credentials if global_credentials
+        config[:region] = if global_configuration
+          config[:region] = region_from_global_configuration || region
```
Comment (DavidS) [inline lib/puppet_x/puppetlabs/aws.rb]: either put the assignment into the if, or the if into the assignment. Please, not both :-)

File: lib/puppet_x/puppetlabs/aws.rb [hunk 2]
```diff
@@ -56,21 +56,85 @@ def self.read_only(*methods)
...
+        else
+          region
         end
         config
       end
 
+      # This method is vendored from the AWS SDK, rather than including an
+      # extra library just to parse an ini file
```
Comment (DavidS) [inline lib/puppet_x/puppetlabs/aws.rb]: Just to make sure: this is not part of the aws-sdk-core gem this currently requires, and adding this as a runtime dependency would require all users to install additional code on nodes running the module?

**Comment by DavidS** [comment]
I only had two minor comments on the code style. Other than that, this should amend the current README. Recommending the IAM role assignment harder and pointing out that `~/.aws` only works with `puppet apply` would be a good start.

**Comment by DavidS** [comment]
While this is still mergable, if you go in an modify the commit, a rebase would also clean up the history.

**Comment by DavidS** [comment]
Merged with slight modifications as #246 .

**Pull request closed without merging**